File: vision/branches/20090415/src/c++/cast/core/CASTWorkingMemory.cpp
/*
 * CAST - The CoSy Architecture Schema Toolkit
 *
 * Copyright (C) 2006-2007 Nick Hawes
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301 USA
 *
 */

#include "CASTWorkingMemory.hpp"
#include <CASTUtils.hpp>


using namespace std;
using namespace cast::cdl;

namespace cast {



  CASTWorkingMemory::CASTWorkingMemory() {

#ifdef SYNC_MEMORY_ACCESS

    pthread_mutexattr_t attr;
    // note: errors here are very unlikely, so we just do a "weaker" overall
    // checking
    int err =  pthread_mutexattr_init(&attr);
    if(err != 0) {
      throw CASTException(exceptionMessage(__HERE__, "failed to create mutex: %s", strerror(err)));
    }


#ifdef __LINUX__    
    //nah: this doesn't work, and isn't necessary, on macs for some reason
    err = pthread_mutexattr_setpshared(&attr, PTHREAD_PROCESS_SHARED);
    if(err != 0) {
      throw CASTException(exceptionMessage(__HERE__, "failed to create mutex: %s", strerror(err)));
    }
#endif

    err = pthread_mutex_init(&m_sync, &attr);
    if(err != 0) {
      throw CASTException(exceptionMessage(__HERE__, "failed to create mutex: %s", strerror(err)));
    }

    err = pthread_mutexattr_destroy(&attr);
    if(err != 0) {
      throw CASTException(exceptionMessage(__HERE__, "failed to create mutex: %s", strerror(err)));
    }

#endif

  }


  /**
   * Empty   destructor.
   */
  CASTWorkingMemory::~CASTWorkingMemory() {
    //TODO memory leak: delete stored objects
  }


  void CASTWorkingMemory::lock()  throw (CASTException) {
#ifdef SYNC_MEMORY_ACCESS
    //lock mutex for wait
    int err = pthread_mutex_lock(&m_sync);
    if(err != 0) {
      throw CASTException(exceptionMessage(__HERE__, "failed mutex lock: %s", strerror(err)));
    }
#endif
  }

  void CASTWorkingMemory::unlock()  throw (CASTException) {
#ifdef SYNC_MEMORY_ACCESS
    int err = pthread_mutex_unlock(&m_sync); 
    if(err != 0) {
      throw CASTException(exceptionMessage(__HERE__, "failed mutex unlock: %s", strerror(err)));
    }
#endif
  }

  /**
   * Adds item to working memory with given id. Does not overwrite
   * data if id already exists.
   * 
   * @param _id
   *            The id of the entry.
   * @param _type
   *            The ontological type of the entry.
   * @param _data
   *            The data itself.
   * @return Returns true if the item is added (i.e. not a duplicate
   *         id)
   */

  bool CASTWorkingMemory::add(const string & _id, 
			      WorkingMemoryEntryPtr _pData) {
  
    assert(_pData->version == 0);

    lock();

    //check whether the id exists
    WMItemMap::iterator i = m_storage.find(_id);
  
    //if(_pData == NULL) {
    //cerr<<"adding null item into wm"<<endl;
    //}
    bool added = false;

    //if the id doesn't exist
    if(i == m_storage.end()) {
    
      //store the data
      m_storage[_id] = _pData;
    
      //store the key in addition order
      m_ids.push_front(_id);

      //     cout<<endl;
      //     cout<<"CASTWorkingMemory data ptr: count: "<<_pData.use_count()<<endl;
      //     cout<<endl;

      //needed for bug #52
      //check to see if we stored it previously
      VersionMap::const_iterator j = m_lastVersions.find(_id);
      if(j != m_lastVersions.end()) {

	//if stored before, new version is last version + 1
	m_storage[_id]->version = (j->second + 1);

	//cout<<"updated stored version from last version: "<<getOverwriteCount(_id)<<endl;
	//erase last version from map
	m_lastVersions.erase(_id);
      }
      
      added = true;
    }

    unlock();

    return added;

  }

  /**
   * Overwrites item with given id. Does not do anything if id does
   * not exist.
   * 
   * @param _id
   *            The id of the entry.
   * @param _type
   *            The ontological type of the entry.
   * @param _data
   *            The data itself.
   * @return Returns true if the id exists for overwriting
   */
  bool CASTWorkingMemory::overwrite(const string &  _id, 
				    WorkingMemoryEntryPtr _pData) {

    lock();

    //check whether the id exists
    WMItemMap::iterator i = m_storage.find(_id);
    bool overwritten = false;

    //if the id exists
    if(i != m_storage.end()) {
    
      //increment overwrite version
      _pData->version = i->second->version + 1;
      //store the data in the same place in the map
      i->second = _pData; 

      //move key to front of list
      m_ids.remove(_id);
      m_ids.push_front(_id);

      overwritten = true;
      //    cout<<"data overwritten"<<endl;
    }

    unlock();

    return overwritten;   
  }

  /**
   * Removes the item with the given id.
   * 
   * @param _id
   *            The id of the entry.
   * @return Returns true if the id is removed.
   */
  WorkingMemoryEntryPtr
  CASTWorkingMemory::remove(const string &  _id) {
  
    lock();

    //check whether the id exists
    WMItemMap::iterator i = m_storage.find(_id);
    WorkingMemoryEntryPtr pItem; //starts NULL
      
    //if the id exists
    if(i != m_storage.end()) {

      pItem = i->second;
    
      //and remove from storage
      m_storage.erase(i);

      //remove key 
      m_ids.remove(_id);

      //store last version
      m_lastVersions[_id] = pItem->version;      
    }

    unlock();

    return pItem;

  }

  /**
   * Get the item with the given id.
   * 
   * @param _id
   *            The specified item or null if it does not exist.
   */
  WorkingMemoryEntryPtr
  CASTWorkingMemory::get(const string &  _id) {

    lock();

    //check whether the id exists
    WMItemMap::iterator i = m_storage.find(_id);
    WorkingMemoryEntryPtr item; //UPGRADE check starts NULL
    
    //if the id exists
    if(i != m_storage.end()) {
      item = i->second;
    }

    //cerr<<"entry not in wm: "<<_id<<endl;
    unlock();
    
    return item;

  }
  

  bool CASTWorkingMemory::contains(const string &  _id) {

    lock();
    //cout<<"contains: "<<_id<<endl;
    //check whether the id exists
    WMItemMap::iterator i = m_storage.find(_id);
    //cout<<"contains: "<<(i != m_storage.end())<<endl;
    bool contains = (i != m_storage.end());

    unlock();

    return contains;
  }

  bool CASTWorkingMemory::hasContained(const string &  _id) {    
 
    lock();

    bool hasContained = false;

    //first check whether we contain it now
    WMItemMap::iterator i = m_storage.find(_id);    
    if(i != m_storage.end()) {
      hasContained = true;
    }
    else {
      //check whether the id exists
      VersionMap::iterator i = m_lastVersions.find(_id);
      hasContained = (i != m_lastVersions.end());
    }

    unlock();

    return hasContained;

  }



  /**
   */
  int CASTWorkingMemory::getOverwriteCount(const string &  _id) {

    lock();

    int count = -1; //TODO smarter default?

    //check whether the id exists
    WMItemMap::iterator i = m_storage.find(_id);
    if(i != m_storage.end()) {
      //cout<<"stored version: "<<_id<<" "<<i->second->version<<endl;
      count =  i->second->version;
    }
    else {
      VersionMap::const_iterator j = m_lastVersions.find(_id);
      if(j != m_lastVersions.end()) {
	//cout<<"last version: "<<_id<<" "<<j->second<<endl;
	count =  j->second;
      }
    }

    //cout<<"never version: "<<_id<<" "<<-1<<endl;
    unlock();

    //TODO make smarter?
    return count;
  }


  /**
   * Gets the n most recent ids of entries with the given type.
   * 
   * @param _type
   *            The type to check.
   * @param _count
   *            The number of ids to return. If 0 all matching items
   *            are return.
   * @return All matching items.
   */
  void
  CASTWorkingMemory::getIDsByType(const string & _type,
				  const int & _count,
				  vector<string> &_ids) {

    lock();  

    int total; 

    if (_count == 0) {
      total = m_storage.size();
    }
    else {
      total = _count;
    }

    int count = 0;

    for(StringList::iterator i = m_ids.begin();
	i != m_ids.end();
	i++) {

      WMItemMap::iterator j = m_storage.find(*i);

      if(j != m_storage.end()) {
	if(j->second->type == _type) {
	  _ids.push_back(*i);
	  count++;
	}

	//will always be at least 1
	if(count == total) {
	  break;
	}
      }
      else {
	assert(false);
      }
    }

    unlock();

  }



  /**
   * Get all working memory entries with given type.
   * 
   * @param _type
   *            The type to check.
   * @return All matching items from working memory.
   */
  void
  CASTWorkingMemory::getByType(const string & _type,
			       vector< WorkingMemoryEntryPtr > & _items) {
    //get all matching items
    getByType(_type,m_storage.size(),_items);
  } 



  /**
   * Get a collection of memory items with the given type.
   * 
   * @param _type
   *            The type to check.
   * @param _count
   *            The number of entries to return. If 0 all matching
   *            entries are returned.
   * @return A collection of matching entries.
   */
  void
  CASTWorkingMemory::getByType(const string & _type, 
			       const int & _count,
			       vector< WorkingMemoryEntryPtr > & _items)  {

    lock();

    int total; 

    if (_count == 0) {
      total = m_storage.size();
    }
    else {
      total = _count;
    }

    int count = 0;

    for(StringList::iterator i = m_ids.begin();
	i != m_ids.end();
	i++) {

      WMItemMap::iterator j = m_storage.find(*i);

      if(j != m_storage.end()) {

	if(j->second->type == _type) {
	  _items.push_back(j->second);
	  //cout<<"returning id: "<<*i<<endl;
	  count++;
	}

	//will always be at least 1
	if(count == total) {
	  break;
	}
      }
      else {
	assert(false);
      }
    }
    unlock();
  }


  void 
  CASTWorkingMemory::debug() {
//     lock();
//     cout<<"pointer counts"<<endl;
//     for(WMItemMap::iterator i = m_storage.begin();
// 	i != m_storage.end();
// 	i++) {
//       cout<<i->first<<" count "<<i->second.use_count()<<endl;
//       //cout<<i->first<<" count "<<i->second->data().use_count()<<endl;

//     }
//    unlock();
  }
}  // namespace cast

File: binding/development/tools/cast/src/java/cast/core/CASTWorkingMemory.java
/*
 * CAST - The CoSy Architecture Schema Toolkit
 *
 * Copyright (C) 2006-2007 Nick Hawes
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301 USA
 *
 */

/**
 * 
 */
package cast.core;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Set;

import cast.cdl.WorkingMemoryEntry;

/**
 * Class that actually operates as a working memory. May eventually implement
 * this in C++ and import via the JNI to ease cross-language development.
 * 
 * @author nah
 */
public class CASTWorkingMemory implements CASTWorkingMemoryInterface {

	// the ontology for the data in working memory. Can be null.
	// private CAATOntology m_ontology;

	// the underlying data storage... a linkedhashmap keeps track of the
	// ordering of the keys so that we can get contents by recency
	private final LinkedHashMap<String, WorkingMemoryEntry> m_storage;

	private final HashMap<String,Integer> m_lastVersions;
	
	/**
	 * Create a new working memory object.
	 */
	public CASTWorkingMemory() {
		m_storage = new LinkedHashMap<String, WorkingMemoryEntry>(20);
		m_lastVersions = new HashMap<String, Integer>(20);
	}

	/**
	 * Get an array of memory items with the given type.
	 * 
	 * @param _type
	 *            The type to check.
	 * @param _count
	 *            The number of entries to return. If 0 all matching entries are
	 *            returned.
	 * @return An array of matching entries.
	 */
	public WorkingMemoryEntry[] getArrayByType(String _type, int _count) {
		Collection<WorkingMemoryEntry> temp = getByType(_type, _count);
		WorkingMemoryEntry[] wmel = new WorkingMemoryEntry[temp.size()];
		temp.toArray(wmel);
		return wmel;
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see caat.core.data.WorkingMemoryInterface#getCollectionByType(java.lang.String,
	 *      int)
	 */
	public Collection<WorkingMemoryEntry> getByType(String _type, int _count) {

		Collection<WorkingMemoryEntry> temp = new ArrayList<WorkingMemoryEntry>();
		int count = 0;

		// need to go backwards through the keys, which is pain!
		Set<String> keys = m_storage.keySet();
		String[] keyArray = new String[keys.size()];
		keys.toArray(keyArray);

		WorkingMemoryEntry item;

		for (int i = keyArray.length - 1; i >= 0; i--) {
			item = m_storage.get(keyArray[i]);
			if (item.type.equals(_type)) {
				temp.add(item);
				count++;
			}
			if (count == _count) {
				break;
			}
		}
		return temp;
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see caat.core.data.WorkingMemoryInterface#getCollectionByType(java.lang.String)
	 */
	public Collection<WorkingMemoryEntry> getByType(String _type) {
		return getByType(_type, m_storage.size());
	}

	/**
	 * Returns a formatted representation of the object. If an ontology is set
	 * then the wm attempts to use this to format the contents of memory.
	 * 
	 * @return The object as in string format.
	 */
	@Override
	public String toString() {
		String out = "[\n";

		Set<String> keys = m_storage.keySet();
		WorkingMemoryEntry item;
		// if (m_ontology == null) {
		for (String id : keys) {
			item = m_storage.get(id);
			// can't do this is final version as data types are
			// unknown!
			out += id + ": [" + item.type + " ["
					+ "<set ontology for printing>" + "]]\n";
		}
		// }
		// else {
		// for (String id : keys) {
		// item = m_storage.get(id);
		// try {
		// // can't do this is final version as data types are
		// // unknown!
		// out += id
		// + ": ["
		// + item.getType()
		// + " ["
		// + RemoteDataTranslator.translateFromAny(item
		// .getData(), m_ontology
		// .ontologicalTypeToDataType(item.getType()))
		// + "]]\n";
		// }
		// catch (FrameworkDataTranslatorException e) {
		// System.err.println(e.getLocalizedMessage());
		// }
		// catch (CAATOntologyException e) {
		// System.err.println(e.getLocalizedMessage());
		// }
		// }
		// }
		out += "\n]";

		return out;
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see caat.core.data.WorkingMemoryInterface#add(java.lang.String,
	 *      java.lang.String, org.omg.CORBA.Any)
	 */
	public boolean add(String _id, WorkingMemoryEntry _data) {

		assert _data.version == 0 : " verion not 0: " + _data.version;

		if (!m_storage.containsKey(_id)) {
			// System.out.println("adding at: " + _id);
			
			if(m_lastVersions.containsKey(_id)) {
//				System.out.println("CASTWorkingMemory.add(): reusing version");
				_data.version = m_lastVersions.get(_id) + 1;
				m_lastVersions.remove(_id);
			}
			
			m_storage.put(_id, _data);
			return true;
		} else {
			return false;
		}
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see caat.core.data.WorkingMemoryInterface#overwrite(java.lang.String,
	 *      java.lang.String, org.omg.CORBA.Any)
	 */
	public boolean overwrite(String _id, WorkingMemoryEntry _data) {
		if (m_storage.containsKey(_id)) {
			// increment overwrite count
			_data.version = m_storage.get(_id).version + 1;
			m_storage.put(_id, _data);
			return true;
		} else {
			System.err.println("data NOT overwritten");
			return false;
		}
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see caat.core.data.WorkingMemoryInterface#remove(java.lang.String)
	 */
	public WorkingMemoryEntry remove(String _id) {
		// System.out.println("looing for : " + _id);
		// System.out.println(this);
		WorkingMemoryEntry item = m_storage.remove(_id);
		if (item == null) {
			return null;
		} else {
			//store version
			m_lastVersions.put(_id, item.version);
			return item;
		}
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see caat.core.data.WorkingMemoryInterface#get(java.lang.String)
	 */
	public WorkingMemoryEntry get(String _id) {
		return m_storage.get(_id);
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see caat.core.data.WorkingMemoryInterface#getIDsOfType(java.lang.String,
	 *      int)
	 */
	public ArrayList<String> getIDsByType(String _type, int _count) {

		if (_count == 0) {
			_count = m_storage.size();
		}

		ArrayList<String> ids = new ArrayList<String>();
		Set<String> keys = m_storage.keySet();
		String[] keyArray = new String[keys.size()];
		keys.toArray(keyArray);

		WorkingMemoryEntry item;

		int count = 0;

		for (int i = keyArray.length - 1; i >= 0; i--) {
			item = m_storage.get(keyArray[i]);
			if (item.type.equals(_type)) {
				ids.add(keyArray[i]);
				count++;
			}
			if (count == _count) {
				break;
			}
		}
		return ids;
	}

	//
	// /*
	// * (non-Javadoc)
	// *
	// * @see caat.core.data.WorkingMemoryInterface#getOntology()
	// */
	// public CAATOntology getOntology() {
	// return m_ontology;
	// }
	//
	// /*
	// * (non-Javadoc)
	// *
	// * @see
	// caat.core.data.WorkingMemoryInterface#setOntology(caat.core.ontologies.CAATOntology)
	// */
	// public void setOntology(CAATOntology _ontology) {
	// m_ontology = _ontology;
	// }

	/*
	 * (non-Javadoc)
	 * 
	 * @see caat.core.data.CAATWorkingMemoryInterface#size()
	 */
	public int size() {
		return m_storage.size();
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see caat.core.data.CAATWorkingMemoryInterface#contains(java.lang.String)
	 */
	public boolean contains(String _queryID) {
		return m_storage.containsKey(_queryID);
	}

	/**
	 * * Whether the wm has ever contained an entry with the given id.
	 * 
	 * @param _queryID
	 * @return
	 */
	public boolean hasContained(String _queryID) {
		if(contains(_queryID)) {
			return true;
		}
		return m_lastVersions.containsKey(_queryID);
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see cast.core.data.CASTWorkingMemoryInterface#getOverwriteCount(java.lang.String)
	 */
	public int getOverwriteCount(String _queryID) {
		WorkingMemoryEntry item = m_storage.get(_queryID);
		if(item != null) {
			return item.version;
		}
		
		Integer lastCount = m_lastVersions.get(_queryID);
		if(lastCount != null) {
			return lastCount;
		}
		
		return -1;
		
		
	}
}
